File: backend/app/api/picks.py
"""Authenticated pick submission + read endpoints.

Thin router mirroring :mod:`app.api.auth` / :mod:`app.api.proof`: it resolves the
acting user from the verified session (NEVER the body), calls the pick-submission
service, and shapes the response. All legality enforcement and the
ViolationCode -> envelope mapping live in :mod:`app.services.pick_submission`; the
router only translates HTTP <-> service and lets the global exception handler emit
the ``{"error": {"code", "message", "reason"}}`` envelope for every rejection.

Security:

* ``user_id`` is always ``user.id`` from :func:`app.api.deps.get_current_user`
  (signed session cookie or bearer) — never from the request body, and there is
  no user parameter on either endpoint, so a user can only read/write their own
  picks (no IDOR). Unauthenticated requests are rejected 401 by the dependency.
* The mutating POST works WITH the existing double-submit CSRF middleware exactly
  as :func:`app.api.proof.echo` does (cookie auth requires ``X-CSRF-Token``;
  bearer is exempt) — there is no CSRF special-casing here.
"""

from __future__ import annotations

from fastapi import APIRouter, Depends
from sqlmodel import Session

from app.api.deps import get_current_user
from app.db import get_session
from app.models import Game, Pick, PickType, Team, User
from app.schemas.picks import PickRead, PickSubmitRequest
from app.services.notifications import (
    COOLDOWN_TTL_SECONDS,
    claim_cooldown,
    misc_picked_event,
    pick_cleared_event,
    pick_event,
    pick_log_detail,
    publish_event,
    roster_complete_event,
)
from app.services.pick_submission import (
    clear_pick,
    main_picks_complete,
    read_picks,
    submit_picks,
)

router = APIRouter(prefix="/api/picks", tags=["picks"])


def _abbr(session: Session, team_id: int | None) -> str | None:
    """Resolve a team id to its abbreviation (display data), or None.

    A tiny read seam used to build the resolved side/team ``detail`` for the
    pickem-logger events. Returns None for a missing/absent id (e.g. a true
    pick'em has no favorite/underdog), which ``pick_log_detail`` tolerates.
    """
    if team_id is None:
        return None
    team = session.get(Team, team_id)
    return team.abbreviation if team is not None else None


def _resolve_pick_detail(session: Session, pick: Pick) -> str:
    """Build the resolved side/team ``detail`` for one persisted pick.

    Loads the pick's Game and its four team abbreviations and delegates to the
    pure :func:`app.services.notifications.pick_log_detail` (which owns the
    favorite/underdog/over/under/MISC + mortal-lock mapping). Display data only.
    """
    game = session.get(Game, pick.game_id)
    if game is None:
        # Should not happen for a just-persisted pick, but never let logging break
        # the request — fall back to the bare pick-type label.
        return pick.pick_type.value
    return pick_log_detail(
        pick.pick_type,
        pick.is_mortal_lock,
        pick.misc_text,
        favorite_abbr=_abbr(session, game.favorite_team_id),
        underdog_abbr=_abbr(session, game.underdog_team_id),
        home_abbr=_abbr(session, game.home_team_id),
        away_abbr=_abbr(session, game.away_team_id),
    )


@router.post("", response_model=list[PickRead])
def submit(
    payload: PickSubmitRequest,
    user: User = Depends(get_current_user),
    session: Session = Depends(get_session),
) -> list[PickRead]:
    """Submit one or more picks for ``{season, week}`` as the current user.

    Window/lock/conflict rejections surface as a structured 4xx envelope (409 for
    window/lock/conflict/roster, 422 for pick'em-spread-ineligible) via the typed
    exceptions the service raises — never a raw 500. On success the picks are
    committed and returned.
    """
    assert user.id is not None  # an authenticated user always has an id
    picks = submit_picks(
        session,
        user_id=user.id,
        season=payload.season,
        week=payload.week,
        items=payload.picks,
    )
    session.commit()
    for pick in picks:
        session.refresh(pick)

    # Post-commit, best-effort pickem-logger publish (one per resulting pick) —
    # AFTER the commit succeeds (caller-owns-commit, mirrors the QT-1 login site).
    # publish_event swallows any Redis error, so a logging hiccup can never break
    # the submit. submit_picks upserts a slot in place, so we cannot cleanly tell
    # an insert from an update from the returned rows — default to ``pick.changed``
    # for the upsert (documented plan choice).
    for pick in picks:
        publish_event(
            pick_event(
                "pick.changed",
                actor=user.display_name,
                week=payload.week,
                detail=_resolve_pick_detail(session, pick),
            )
        )

        # QT-3 pickem-CHAT (260628-itg): when THIS submit set/updated a MISC
        # prediction, fire ONE leak-safe misc.picked line announcing only THAT
        # the player made their misc call — never the misc_text. Gated behind a
        # ~5-min cooldown per (user, season, week) so repeated edits to the
        # prediction within the window do not re-spam the chat feed. Done AFTER
        # the pick.changed publish so the logger event is unaffected; the
        # cooldown is fail-open and publish_event swallows, so it stays
        # post-commit + best-effort.
        if pick.pick_type is PickType.MISC:
            misc_key = (
                f"pickem:misc_picked_cd:{payload.season}:{payload.week}:{user.id}"
            )
            if claim_cooldown(misc_key, COOLDOWN_TTL_SECONDS):
                publish_event(
                    misc_picked_event(actor=user.display_name, week=payload.week)
                )

    # QT-3 pickem-CHAT: when THIS submit results in the user holding their full
    # standard card for the week — all four base bet types plus a mortal lock —
    # fire ONE roster.complete (display_name only) to the chat feed — post-commit
    # + best-effort (publish_event swallows). Adding the mortal lock can now be
    # the submit that completes the card (no longer a no-op for completion). Gated
    # behind a ~5-min cooldown per (user, season, week) so an immediate second
    # completing submit (e.g. re-setting a slot while the card stays complete)
    # does NOT re-post the roster line (260628-itg). A submit that leaves the card
    # incomplete fires none; the cooldown is fail-open so a Redis hiccup can never
    # suppress the milestone.
    if main_picks_complete(
        session, user_id=user.id, season=payload.season, week=payload.week
    ):
        roster_key = (
            f"pickem:roster_complete_cd:{payload.season}:{payload.week}:{user.id}"
        )
        if claim_cooldown(roster_key, COOLDOWN_TTL_SECONDS):
            publish_event(
                roster_complete_event(actor=user.display_name, week=payload.week)
            )
    return [PickRead.from_orm_pick(p) for p in picks]


@router.get("", response_model=list[PickRead])
def read(
    season: int,
    week: int,
    user: User = Depends(get_current_user),
    session: Session = Depends(get_session),
) -> list[PickRead]:
    """Read the current user's own picks for ``{season, week}``.

    Scoped to the session user by construction — no user parameter is accepted,
    so reading another user's picks is impossible.
    """
    assert user.id is not None
    picks = read_picks(session, user_id=user.id, season=season, week=week)
    return [PickRead.from_orm_pick(p) for p in picks]


@router.delete("", status_code=204)
def clear(
    season: int,
    week: int,
    pick_type: PickType,
    is_mortal_lock: bool = False,
    user: User = Depends(get_current_user),
    session: Session = Depends(get_session),
) -> None:
    """Clear (delete) the current user's pick for one ``{season, week, pick_type,
    is_mortal_lock}`` slot — the un-pick counterpart of POST.

    Security / user-scoping: the acting user is always ``user.id`` from
    :func:`app.api.deps.get_current_user` (signed session cookie or bearer) — there
    is NO user parameter in the path/query/body, and :func:`clear_pick` filters by
    ``user_id``, so a user can only ever delete their OWN pick (no IDOR).
    Unauthenticated requests are rejected 401 by the dependency.

    Enforcement: the SAME pick-window / per-game-lock rules as POST apply, delegated
    entirely to :func:`clear_pick` (no duplicated window math). A closed window or a
    kicked-off game surfaces as a 409 envelope; a missing slot as a 404
    (``pick_not_found``) — and nothing is deleted on any rejection.

    Shape choice (D-01): the slot identifiers are QUERY PARAMETERS, not a request
    body. DELETE-with-body is fragile across the httpx TestClient / proxies and
    FastAPI treats a DELETE body as non-idiomatic, whereas query params are the
    idiomatic REST shape for a DELETE and keep the auth/CSRF surface identical to the
    other endpoints. The slot fields mirror the POST contract exactly (``PickType``
    enum value for ``pick_type``, bool for ``is_mortal_lock``; ``is_mortal_lock``
    defaults to ``False`` so clearing a base slot needs only season/week/pick_type).

    The mutating DELETE works WITH the existing double-submit CSRF middleware exactly
    as POST does (cookie auth requires ``X-CSRF-Token``; bearer is exempt) — there is
    no CSRF special-casing here. On success FastAPI emits 204 No Content (no body).
    """
    assert user.id is not None  # an authenticated user always has an id
    clear_pick(
        session,
        user_id=user.id,
        season=season,
        week=week,
        pick_type=pick_type,
        is_mortal_lock=is_mortal_lock,
    )
    session.commit()

    # Post-commit, best-effort pickem-logger publish. The row is already deleted,
    # so we have no Game to resolve a team abbreviation against — name the cleared
    # SLOT from the pick_type/lock the caller passed (pick_log_detail falls back to
    # the bare side label, e.g. "OVER" / "Favorite", when abbrs are absent).
    detail = pick_log_detail(
        pick_type,
        is_mortal_lock,
        None,
        favorite_abbr=None,
        underdog_abbr=None,
        home_abbr=None,
        away_abbr=None,
    )
    publish_event(
        pick_cleared_event(actor=user.display_name, week=week, detail=detail)
    )
    return None

File: frontend/src/lib/picks.ts
/**
 * Typed picks/slate API layer for the My Picks page.
 *
 * Mirrors the small, one-function-per-endpoint style of lib/currentWeek.ts and
 * lib/config.ts: every call routes through the CSRF-aware `api<T>()` wrapper
 * (credentials:include + X-CSRF-Token on unsafe methods) — never a raw fetch.
 *
 * Wire shapes match the backend exactly (see backend/app/schemas/slate.py and
 * backend/app/schemas/picks.py). NOTE: Decimal line fields (spread/total)
 * serialize to JSON as strings, so they are typed `string | null` here.
 */
import { api } from "./api";

/**
 * Pick types (literal union mirroring backend PickType). The first four are the
 * BASE bet types (spread/total, each with a slate eligibility + a favorite/
 * underdog or over/under side). MISC is a fifth, NON-base type: a weekly
 * free-text prediction tied to any game — it has NO slate eligibility (the slate
 * still ships an `eligibility["MISC"]` key, but it is meaningless and must never
 * be indexed) and NO favorite/underdog/over-under side. MISC is never a mortal
 * lock and is never rendered as a per-game base button.
 */
export type PickType =
  | "UNDERDOG_COVER"
  | "FAVORITE_COVER"
  | "OVER"
  | "UNDER"
  | "MISC";

/** Grading state of a persisted pick (mirrors backend PickResult). */
export type PickResult = "PENDING" | "WIN" | "LOSS";

/** Public team reference identity for one side of a matchup. */
export interface SlateTeam {
  team_id: number;
  abbreviation: string;
  display_name: string;
}

/**
 * One game's pickable slate entry: identity, line, lock + per-type eligibility.
 * `spread`/`total` are JSON strings (backend Decimal) — keep them as strings.
 */
export interface SlateGame {
  game_id: number;
  kickoff_at: string | null;
  home_team: SlateTeam;
  away_team: SlateTeam;
  spread: string | null;
  total: string | null;
  favorite_team_id: number | null;
  underdog_team_id: number | null;
  locked: boolean;
  eligibility: Record<PickType, boolean>;
}

/** The pickable slate for a {season, week}: one entry per game. */
export interface Slate {
  season: number;
  week: number;
  games: SlateGame[];
  /**
   * The COMPUTED week-level odds-freeze flag (mirrors backend
   * SlateResponse.odds_frozen) — true once this week's lines are frozen. Used to
   * gate the My Picks re-pick notice for now-ineligible held picks.
   */
  odds_frozen: boolean;
}

/** A persisted pick as returned to its owner (never exposes user_id). */
export interface PickRead {
  id: number;
  game_id: number;
  week_id: number;
  pick_type: PickType;
  is_mortal_lock: boolean;
  result: PickResult;
  points: number;
  /**
   * The free-text prediction for a MISC pick (mirrors backend PickRead.misc_text).
   * The owner ALWAYS sees their own misc_text; it is NULL/absent for every
   * non-MISC pick type.
   */
  misc_text?: string | null;
}

/** A single autosave item (one pick on one game). */
export interface PickItem {
  game_id: number;
  pick_type: PickType;
  is_mortal_lock: boolean;
  /**
   * The free-text prediction (mirrors backend PickItem.misc_text). Sent ONLY on
   * a MISC pick (required, non-blank server-side via `misc_text_required`);
   * omitted for base picks (the backend rejects `misc_text` on a non-MISC pick
   * via `misc_text_not_allowed`, so the UI never sends it for base types).
   */
  misc_text?: string | null;
}

/**
 * Stable string key for a pick *slot*: `(pick_type, is_mortal_lock)`. The four
 * base slots are is_mortal_lock=false (one per pick_type); the mortal lock is
 * its own slot (is_mortal_lock=true) and may duplicate a base pick_type. Used to
 * merge returned PickRead rows into the picks map.
 */
export function slotKey(pick_type: PickType, is_mortal_lock: boolean): string {
  return `${pick_type}|${is_mortal_lock}`;
}

/**
 * Stable key for an inline pick *error*, scoped to the specific GAME + slot the
 * user acted on — `(game_id, pick_type, is_mortal_lock)`. Unlike slotKey, this
 * includes game_id so a rejection (e.g. a contradiction) renders only on the
 * control that was clicked, not on every same-type button across cards.
 */
export function errorKey(
  game_id: number,
  pick_type: PickType,
  is_mortal_lock: boolean,
): string {
  return `${game_id}|${slotKey(pick_type, is_mortal_lock)}`;
}

/** Fetch the pickable slate (the OPTIONS the page renders) for a week. */
export function getSlate(season: number, week: number): Promise<Slate> {
  return api<Slate>(`/api/slate?season=${season}&week=${week}`);
}

/** Fetch the current user's authoritative roster for a week. */
export function getMyPicks(season: number, week: number): Promise<PickRead[]> {
  return api<PickRead[]>(`/api/picks?season=${season}&week=${week}`);
}

/**
 * Autosave a SINGLE pick item. The endpoint accepts an array, but the page
 * commits one pick at a time (each selection fires its own POST). Returns only
 * the AFFECTED rows (not the whole roster). Content-Type + X-CSRF-Token are
 * handled by api().
 */
export function submitPick(
  season: number,
  week: number,
  item: PickItem,
): Promise<PickRead[]> {
  return api<PickRead[]>("/api/picks", {
    method: "POST",
    body: JSON.stringify({ season, week, picks: [item] }),
  });
}

/**
 * Clear (un-pick) a SINGLE slot via DELETE /api/picks. Unlike submitPick, the
 * slot identifiers are QUERY PARAMS (not a body): season, week, pick_type, and
 * is_mortal_lock (booleans serialized as the literal strings `true`/`false`).
 * The endpoint returns 204 (no body) on success, so the result is void (api()
 * returns undefined for 204). X-CSRF-Token is attached by api() on DELETE.
 */
export function clearPick(
  season: number,
  week: number,
  slot: { pick_type: PickType; is_mortal_lock: boolean },
): Promise<void> {
  const query =
    `season=${season}&week=${week}` +
    `&pick_type=${slot.pick_type}&is_mortal_lock=${slot.is_mortal_lock}`;
  return api<void>(`/api/picks?${query}`, { method: "DELETE" });
}
